tempesta_fw/t/unit/sse_parser/sse_parser.c

```diff
@@ -0,0 +1,1552 @@
+    c->swapbytes16 = _mm_set_epi8(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15);
+    c->cvt10_2_bytes = _mm_set1_epi16(256+10);
+    c->cvt10_2_words = _mm_set1_epi32(65536+100);
+
+    c->spaces = _mm_set1_epi8(0x20);
+    c->semicolons = _mm_set1_epi8(':');
+    c->slashes = _mm_set1_epi8('/');
+    c->tabs = _mm_set1_epi8('\t');
+    c->plus = _mm_set1_epi8('+');
+    c->nine = _mm_set1_epi8('9');
+    c->zero = _mm_set1_epi8('0');
+    c->_n   = _mm_set1_epi8('\n');
+    c->_r   = _mm_set1_epi8('\r');
+
+    c->tabs2spaces = _mm_set1_epi8(' '-'\t');
+    c->plus2spaces = _mm_set1_epi8('+'-' ');
+
+    c->masktest1 = _mm_setr_epi8(1,2,4,8,16,32,64,128,0,0,0,0,0,0,0,0);
+    c->masktest2 = _mm_set1_epi8(0xF);
+
+    c->newline1 = _mm_setr_epi8('\n','\n', 0  , 0  ,
+                                '\r','\n','\n', 0  ,
+                                '\n','\r','\n', 0  ,
+                                '\r','\n','\r','\n');
+    c->newline2 = _mm_setr_epi8(0xFF,0xFF, 0  , 0  ,
+                                0xFF,0xFF,0xFF, 0  ,
+                                0xFF,0xFF,0xFF, 0  ,
+                                0xFF,0xFF,0xFF,0xFF);
+    c->dq0 = _mm_setr_epi8(
+                0x01, 0x01, 0x01, 0x01, 0x02, 0x02, 0x02, 0x02,
+                0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
+    c->dq1 = _mm_setr_epi8(
+                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20,
+                0x00, 0x00, 0x20, 0x20, 0x00, 0x00, 0x20, 0x20);
+    c->dq2 = _mm_setr_epi8(
+                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66,
+                0x2F, 0x39, 0x60, 0x66, 0x2F, 0x39, 0x60, 0x66);
+    c->dq3 = _mm_set1_epi16(0x00FF);
+    c->dq4 = _mm_setr_epi8(
+                0x30, 0xFF, 0x57, 0xFF, 0x30, 0xFF, 0x57, 0xFF,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00);
+    c->dq5 = _mm_setr_epi16(
+                16, 16, 1, 1, 0, 0, 0, 0);
+
+    c->directoryPatterns =
+            _mm_setr_epi8('/', '.', '.', '/',
+                          '/', '.', '/',  0 ,
+                          '/', '/',  0 ,  0 ,
+                           0 ,  0 ,  0 ,  0 );
+
+    static const char * http_m[] = {
+        "GET", "PUT", "POST", "COPY", "MOVE", "LOCK", "HEAD", "PATCH",
+        "TRACK", "DELETE", "UNLOCK", "MKCOL", "OPTIONS", "PROPFIND", "PROPPATCH",
+        NULL
+    };
+    XASSERT(sizeof(c->http_method) >= tokenSetLength(http_m));
+    initTokenSet(http_m, &c->http_method, sizeof(c->http_method));
+
+    static const char * http_sch[] = {
+        "https://", "http://", NULL
+    };
+    XASSERT(sizeof(c->http_schema) >= tokenSetLength(http_sch));
+    initTokenSet(http_sch, &c->http_schema, sizeof(c->http_schema));
+
+    static const char * http_v[] = {
+        "\n", "\r\n", "HTTP/1.0\n", "HTTP/1.0\r\n",
+        "HTTP/1.1\n", "HTTP/1.1\r\n", NULL
+    };
+    XASSERT(sizeof(c->http_version) >= tokenSetLength(http_v));
+    initTokenSet(http_v, &c->http_version, sizeof(c->http_version));
+
```

Comment:
Well, `http_m`, `http_sch` and `http_v` contains very short strings,
so there is no sense to use vector processing. Your becnhmarks
implicitly show this: GOTO outperforms SSE-ff on short strings.
Also keep in ming that `GET / HTTP/1.0\n\n` is a valid request which
we can face on HTTP Flood and we need maximum performance in this case
as well as in normal case with long URIs and Cookies.
Therefore, resulting parser must combine both the approaches and use
vectorized processing for long strings, i.e. URIs and headers. Probably,
you find a heuristic whether to use vectorized parser processing.
I'd imagine it could be length of rest of ingress `data`.
Typically HTTP messages don't contain many spaces sequently, so there is
also no need to use SSE for the processing. However, it'd be good TODO
for Frang to catch such divitating messages... For now we can cath such
requests by limiting size of HTTP fields.